File: scripts/unreal/walking.mjs
import assert from "node:assert/strict";
import { createHash } from "node:crypto";
import { mkdir, readFile, rename, rm, writeFile } from "node:fs/promises";
import { dirname, isAbsolute, relative, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import ts from "typescript";

const ROOT = resolve(dirname(fileURLToPath(import.meta.url)), "../..");
const VIEWPORT = "lib/twin-viewport-contract.ts";
const CONSTANTS = ["WALK_EYE_HEIGHT_M", "WALK_SPEED_MPS", "WALK_CAMERA", "WALK_SURFACE",
  "WALK_COLLISION_ELLIPSOID_M", "WALK_COLLISION_OFFSET_M"];
const sha = (bytes) => createHash("sha256").update(bytes).digest("hex");
const cm = (metres) => Number((metres * 100).toPrecision(15));
const mm = (metres) => Number((metres * 1000).toPrecision(15));
const finite = (value, label) => assert(Number.isFinite(value), `${label} must be finite`);

/** Read only literal declarations; never execute the renderer or evaluate source code. */
export function readWalkingConstants(source) {
  const file = ts.createSourceFile(VIEWPORT, source, ts.ScriptTarget.Latest, true);
  assert.equal(file.parseDiagnostics.length, 0, "Invalid viewport TypeScript");
  function literal(node) {
    if (ts.isNumericLiteral(node)) return Number(node.text);
    if (ts.isPrefixUnaryExpression(node) && node.operator === ts.SyntaxKind.MinusToken) return -literal(node.operand);
    if (ts.isObjectLiteralExpression(node)) {
      return Object.fromEntries(node.properties.map((property) => {
        assert(ts.isPropertyAssignment(property) && ts.isIdentifier(property.name), "Expected named numeric property");
        return [property.name.text, literal(property.initializer)];
      }));
    }
    if (ts.isCallExpression(node) && node.expression.getText(file) === "Object.freeze" && node.arguments.length === 1) {
      return literal(node.arguments[0]);
    }
    throw new Error(`Unsupported walking constant initializer: ${node.getText(file)}`);
  }
  const result = {};
  for (const statement of file.statements) {
    if (!ts.isVariableStatement(statement)) continue;
    for (const declaration of statement.declarationList.declarations) {
      const name = declaration.name.getText(file);
      if (!CONSTANTS.includes(name)) continue;
      assert(!(name in result), `Duplicate walking constant: ${name}`);
      assert(declaration.initializer, `Missing initializer: ${name}`);
      result[name] = literal(declaration.initializer);
    }
  }
  assert.deepEqual(Object.keys(result).sort(), [...CONSTANTS].sort(), "Missing walking source constants");
  return result;
}

/** OBJ is right handed Z-up mm; the native bridge flips plan Y and scales to cm. */
export function walkingBoundsToUnreal(bounds) {
  for (const key of ["min", "max"]) {
    assert(Array.isArray(bounds?.[key]) && bounds[key].length === 3, `Invalid bounds.${key}`);
    bounds[key].forEach((n) => finite(n, `bounds.${key}`));
  }
  bounds.min.forEach((n, i) => assert(n <= bounds.max[i], "Inverted source bounds"));
  return {
    min: [bounds.min[0] / 10, -bounds.max[1] / 10, bounds.min[2] / 10],
    max: [bounds.max[0] / 10, -bounds.min[1] / 10, bounds.max[2] / 10],
  };
}

// This is the exact render/archive rule in convert.py and validate.mjs.
function runtimeObject(object) {
  return object.enabled && !["odlesk vod", "kaustick", "odraz oblohy"].some((part) => object.name.toLowerCase().includes(part));
}

function reference(object) {
  assert(typeof object.sourceId === "string" && object.sourceId.length, `${object.id}: missing sourceId`);
  assert(Number.isSafeInteger(object.triangles) && object.triangles > 0, `${object.id}: invalid source triangles`);
  assert(Number.isSafeInteger(object.instances) && object.instances > 0, `${object.id}: invalid source instances`);
  return {
    objectId: object.id,
    sourceId: object.sourceId,
    entityId: object.metadata?.entityId ?? null,
    name: object.name,
    sourceBoundsMm: structuredClone(object.boundsMm),
    nativeBoundsCm: walkingBoundsToUnreal(object.boundsMm),
    sourceGeometry: { objectId: object.id, triangles: object.triangles, instances: object.instances },
  };
}

/** Pure manifest builder. It exports references to captured triangles, never new geometry. */
export function buildWalkingContract(scene, constants, provenance, { nativeCapsuleHalfHeightCm = 85 } = {}) {
  assert.equal(scene.units, "millimetres", "Walking requires the millimetre scene manifest");
  assert.equal(scene.coordinateSystem, "right-handed Z-up", "Unexpected source coordinate system");
  assert(Array.isArray(scene.objects), "Missing source objects");
  finite(scene.sceneCenterMm?.x, "sceneCenterMm.x");
  finite(scene.sceneCenterMm?.y, "sceneCenterMm.y");
  for (const key of ["sceneSha256", "viewportSha256", "writerSha256"]) {
    assert(/^[a-f0-9]{64}$/.test(provenance?.[key] ?? ""), `Missing ${key}`);
  }
  const source = structuredClone(constants);
  for (const name of CONSTANTS) {
    assert(name in source, `Missing ${name}`);
    for (const value of typeof source[name] === "object" ? Object.values(source[name]) : [source[name]]) finite(value, name);
  }
  const ellipsoid = source.WALK_COLLISION_ELLIPSOID_M;
  const offset = source.WALK_COLLISION_OFFSET_M;
  assert(ellipsoid.x > 0 && ellipsoid.y > 0 && ellipsoid.z > 0, "Invalid source ellipsoid");
  assert.equal(ellipsoid.x, ellipsoid.z, "Asymmetric ellipsoid needs a reviewed native capsule adaptation");
  assert.equal(offset.x, 0, "Nonvertical collider offset needs native review");
  assert.equal(offset.z, 0, "Nonvertical collider offset needs native review");
  const eyeHeightCm = cm(source.WALK_EYE_HEIGHT_M);
  const capsuleRadiusCm = cm(ellipsoid.x);
  finite(nativeCapsuleHalfHeightCm, "nativeCapsuleHalfHeightCm");
  assert(nativeCapsuleHalfHeightCm >= capsuleRadiusCm && nativeCapsuleHalfHeightCm * 2 > eyeHeightCm,
    "Native capsule must enclose the eye with positive head clearance");
  assert(eyeHeightCm > 0 && source.WALK_SURFACE.maxStepUpM > 0 && source.WALK_SURFACE.maxDropM > 0,
    "Invalid source movement distances");
  assert(Object.values(source.WALK_SPEED_MPS).every((n) => n > 0), "Invalid source walking speed");
  assert.deepEqual(Object.keys(source.WALK_SPEED_MPS).sort(), ["boost", "normal", "precision"]);

  const ids = new Set();
  const walkSurfaceIds = new Set();
  const walkSurfaces = [], staticBlockers = [], capturedClosedBlockers = [];
  let excludedDynamicMembers = 0, archivedObjects = 0;
  for (const object of scene.objects) {
    assert(typeof object.id === "string" && /^DOM_\d+$/.test(object.id), "Invalid canonical object ID");
    assert(!ids.has(object.id), `Duplicate source object ID: ${object.id}`);
    ids.add(object.id);
    assert.equal(typeof object.enabled, "boolean", `${object.id}: missing enabled state`);
    assert.equal(typeof object.name, "string", `${object.id}: missing source name`);
    if (!runtimeObject(object)) { archivedObjects += 1; continue; }
    const metadata = object.metadata ?? {};
    assert.equal(typeof metadata.babylonCheckCollisions, "boolean",
      `${object.id}: missing captured babylonCheckCollisions; re-export the source scene`);
    for (const key of ["walkSurface", "cameraOccluder", "dynamicCameraOccluder"]) {
      assert(metadata[key] === undefined || typeof metadata[key] === "boolean", `${object.id}: invalid ${key}`);
    }
    const dynamic = metadata.dynamicCameraOccluder === true || Boolean(metadata.doorMotion);
    const collision = metadata.babylonCheckCollisions || metadata.walkSurface === true || metadata.cameraOccluder === true;
    if (dynamic && !collision) { excludedDynamicMembers += 1; continue; }
    if (!collision) continue;
    const ref = { ...reference(object), collisionReasons: [
      ...(metadata.babylonCheckCollisions ? ["babylon-check-collisions"] : []),
      ...(metadata.walkSurface === true ? ["explicit-walk-surface"] : []),
      ...(metadata.cameraOccluder === true ? ["explicit-camera-occluder"] : []),
    ] };
    if (dynamic) {
      assert(typeof metadata.doorId === "string" && metadata.doorId.length, `${object.id}: dynamic blocker lacks doorId`);
      assert(metadata.doorMotion === undefined || ["HINGED", "SLIDING", "OVERHEAD"].includes(metadata.doorMotion),
        `${object.id}: unreviewed door motion`);
      capturedClosedBlockers.push({ ...ref, doorId: metadata.doorId, motion: metadata.doorMotion ?? null,
        subject: metadata.doorSubject ?? null, cameraOccluder: metadata.cameraOccluder === true,
        walkSurface: metadata.walkSurface === true, requiredState: "captured-closed", collisionRequired: true,
        runtimeTags: ["BreziClosedBlocker", `BreziSourceObjectId=${object.id}`, `BreziSourceId=${object.sourceId}`, `BreziDoorId=${metadata.doorId}`] });
    } else {
      staticBlockers.push({ ...ref, cameraOccluder: metadata.cameraOccluder === true,
        walkSurface: metadata.walkSurface === true, babylonCheckCollisions: metadata.babylonCheckCollisions, collisionRequired: true });
    }
    if (metadata.walkSurface === true) {
      assert(typeof metadata.walkSurfaceId === "string" && metadata.walkSurfaceId.length, `${object.id}: missing walkSurfaceId`);
      assert(!walkSurfaceIds.has(metadata.walkSurfaceId), `Duplicate walkSurfaceId: ${metadata.walkSurfaceId}`);
      walkSurfaceIds.add(metadata.walkSurfaceId);
      assert(["interior", "exterior", "terrain"].includes(metadata.walkSurfaceKind), `${object.id}: missing walkSurfaceKind`);
      finite(metadata.walkSurfaceElevationOffsetM, `${object.id}: walkSurfaceElevationOffsetM`);
      const supportOffsetCm = cm(metadata.walkSurfaceElevationOffsetM);
      walkSurfaces.push({ ...ref, walkSurfaceId: metadata.walkSurfaceId, kind: metadata.walkSurfaceKind,
        supportOffsetCm, sourceElevationOffsetM: metadata.walkSurfaceElevationOffsetM,
        sourceElevationOffsetMm: mm(metadata.walkSurfaceElevationOffsetM),
        supportTranslationCm: [0, 0, supportOffsetCm],
        dynamic, doorId: metadata.doorId ?? null,
        query: "source-triangle-hit-plus-authored-vertical-offset",
        runtimeTags: ["BreziWalkSurface", `BreziSourceObjectId=${object.id}`, `BreziSourceId=${object.sourceId}`,
          `BreziWalkSurfaceId=${metadata.walkSurfaceId}`],
        supportRuntimeTags: supportOffsetCm === 0 ? [] : ["BreziWalkSupportProxy", `BreziSupportOffsetCm=${supportOffsetCm}`] });
    }
  }
  assert(walkSurfaces.length > 0, "No source walk surfaces");
  return {
    schemaVersion: 1,
    status: "source-walking-contract-exported",
    coordinateSystem: "unreal-centimeters",
    sourceCoordinateSystem: "right-handed Z-up millimetres",
    coordinateTransform: "source OBJ [x,y,z] mm -> Unreal [x/10,-y/10,z/10] cm",
    sceneCenterMm: structuredClone(scene.sceneCenterMm),
    layoutId: scene.layoutId,
    provenance: { ...provenance, viewportSource: VIEWPORT, sourceObjSha256: scene.objSha256 ?? null },
    eyeHeightCm, capsuleRadiusCm, capsuleHalfHeightCm: nativeCapsuleHalfHeightCm,
    speedsCmPerSecond: Object.fromEntries(Object.entries(source.WALK_SPEED_MPS).map(([key, value]) => [key, cm(value)])),
    maxStepHeightCm: cm(source.WALK_SURFACE.maxStepUpM), maxDropCm: cm(source.WALK_SURFACE.maxDropM),
    probeHeadroomCm: cm(source.WALK_SURFACE.probeHeadroomM),
    sourceCollider: {
      kind: "ellipsoid", ellipsoidRadiiCm: [cm(ellipsoid.x), cm(ellipsoid.z), cm(ellipsoid.y)],
      centerOffsetCm: [cm(offset.x), cm(offset.z), cm(offset.y)],
      topAboveFloorCm: cm(offset.y) + cm(ellipsoid.y), bottomAboveFloorCm: cm(offset.y) - cm(ellipsoid.y),
    },
    nativeAdaptation: {
      id: "NATIVE-CHARACTER-CAPSULE-20260908", kind: "capsule-approximation-of-source-ellipsoid",
      reason: `ACharacter sweeps a capsule. Its floor-based height includes the ${eyeHeightCm} cm source eye; the source ellipsoid ends at ${cm(offset.y + ellipsoid.y)} cm.`,
      sourceHalfHeightCm: cm(ellipsoid.y), sourceCenterHeightCm: cm(offset.y),
      nominalCenterHeightCm: nativeCapsuleHalfHeightCm, eyeOffsetFromCenterCm: eyeHeightCm - nativeCapsuleHalfHeightCm,
      floorClearancePolicy: "engine-floor-gap-compensated-in-camera",
      nativeHeadClearanceCm: 2 * nativeCapsuleHalfHeightCm - eyeHeightCm, geometricallyIdentical: false,
    },
    sourceConstants: source,
    walkSurfaces, staticBlockers, capturedClosedBlockers,
    summary: { sourceObjects: scene.objects.length, archivedObjects, walkSurfaces: walkSurfaces.length,
      offsetSurfaces: walkSurfaces.filter((s) => s.supportOffsetCm !== 0).length, staticBlockers: staticBlockers.length,
      capturedClosedBlockers: capturedClosedBlockers.length,
      capturedCheckCollisionObjects: scene.objects.filter((o) => runtimeObject(o) && o.metadata?.babylonCheckCollisions).length,
      closedDoorOrHatchGroups: new Set(capturedClosedBlockers.map((b) => b.doorId)).size, excludedDynamicMembers },
    collisionContract: {
      geometry: "existing-runtime-GLB-source-triangles", boundsAreBroadPhaseOnly: true,
      blockerEligibility: "captured-babylon-checkCollisions-or-explicit-walkSurface-or-cameraOccluder",
      skippedSourceColliders: structuredClone((scene.skipped ?? []).filter((r) => r.enabled && r.babylonCheckCollisions)),
      visualMeshesMustRemainUnchanged: true,
      offsetsApplyTo: "support-query-hit-height-or-exact-translated-source-triangles",
      floorEligibility: "only-explicit-walkSurfaces", closedBlockerPolicy: "preserve-captured-transforms-and-enable-collision",
      doorAnimationExported: false, doorProgressSerializedByScene: false,
      closedStateEvidence: "fresh-scene capture; AnimatedDoorController defaults initiallyOpen to false; hatchProgress starts at zero",
    },
    nativeValidation: { characterMovementImplemented: false, collisionQueriesVerified: false, runtimeWalkingVerified: false },
    pending: [
      "Bind source object IDs to existing native mesh/component assets and retain serialized runtime tags.",
      "Configure ACharacter capsule, eye offset, speeds, step and source surface/drop policy; verify swept collision and low ceilings.",
      "Apply every authored support offset to queries or exact source-mesh support components; do not use AABBs as floors.",
      "Enable every captured closed door/hatch blocker; animated door progress/transforms require a separate source export and native implementation.",
      "Review skipped hidden source colliders separately; this contract does not claim they were exported as render meshes.",
      "Verify floor eligibility, thresholds, deck gaps, pool/shaft voids and transitions in the native runtime before claiming walk support.",
    ],
  };
}

/** File API for the export pipeline; all captured source hashes must still match. */
export async function exportWalking({ rootDir = ROOT, scenePath = resolve(rootDir, "output/unreal/geometry/scene.json"),
  outputPath = resolve(dirname(scenePath), "walking.json"), nativeCapsuleHalfHeightCm = 85 } = {}) {
  const bytes = await readFile(scenePath), scene = JSON.parse(bytes);
  assert(scene.sourceFiles && typeof scene.sourceFiles === "object", "Missing source hashes");
  assert(scene.sourceFiles[VIEWPORT], "Scene lacks the viewport source checksum");
  for (const [name, digest] of Object.entries(scene.sourceFiles)) {
    const path = resolve(rootDir, name), rel = relative(rootDir, path);
    assert(!isAbsolute(rel) && rel !== ".." && !rel.startsWith("../"), `Source path escapes project: ${name}`);
    assert.equal(sha(await readFile(path)), digest, `Source changed since scene capture: ${name}`);
  }
  assert.equal(sha(await readFile(resolve(dirname(scenePath), "dom-mm.obj"))), scene.objSha256, "Source OBJ checksum mismatch");
  const viewport = await readFile(resolve(rootDir, VIEWPORT));
  const writer = await readFile(fileURLToPath(import.meta.url));
  const contract = buildWalkingContract(scene, readWalkingConstants(viewport.toString("utf8")), {
    sceneSha256: sha(bytes), viewportSha256: sha(viewport), writerSha256: sha(writer),
  }, { nativeCapsuleHalfHeightCm });
  // Guard the validation-to-write window; never bless an export from changed bytes.
  assert.equal(sha(await readFile(scenePath)), contract.provenance.sceneSha256, "Scene changed during walking export");
  const output = `${JSON.stringify(contract, null, 2)}\n`, temporary = `${outputPath}.${process.pid}.tmp`;
  await mkdir(dirname(outputPath), { recursive: true });
  try { await writeFile(temporary, output); await rename(temporary, outputPath); }
  finally { await rm(temporary, { force: true }); }
  return { outputPath, sha256: sha(output), contract };
}

if (process.argv[1] && resolve(process.argv[1]) === fileURLToPath(import.meta.url)) {
  const directory = resolve(ROOT, process.env.UNREAL_OUTPUT ?? "output/unreal/geometry");
  const result = await exportWalking({ scenePath: resolve(directory, "scene.json"),
    outputPath: resolve(process.env.UNREAL_WALKING_OUTPUT ?? resolve(directory, "walking.json")) });
  console.log(JSON.stringify({ status: result.contract.status, outputPath: result.outputPath, sha256: result.sha256,
    summary: result.contract.summary, nativeValidation: result.contract.nativeValidation }, null, 2));
}

File: scripts/unreal/walking.py
"""Author source-triangle walking collision; runtime CharacterMovement is separate."""
import hashlib
import json
from pathlib import Path

import unreal

ROOT = Path(__file__).resolve().parents[2]
OWNER = "scripts/unreal/walking.py"


def sha(path):
    return hashlib.sha256(path.read_bytes()).hexdigest()


def add_tags(obj, values):
    current = list(obj.get_editor_property("tags" if isinstance(obj, unreal.Actor) else "component_tags"))
    names = {str(value) for value in current}
    current.extend(unreal.Name(value) for value in values if value not in names)
    obj.set_editor_property("tags" if isinstance(obj, unreal.Actor) else "component_tags", current)


def bounds(component):
    origin, extent, _ = unreal.SystemLibrary.get_component_bounds(component)
    return {"min": [getattr(origin, axis) - getattr(extent, axis) for axis in ("x", "y", "z")],
            "max": [getattr(origin, axis) + getattr(extent, axis) for axis in ("x", "y", "z")]}


def apply_walking_contract(scene, meshes_by_id, geometry_dir):
    from materials import _asset_hashes

    geometry_dir = Path(geometry_dir)
    contract_file = geometry_dir / "walking.json"
    contract = json.loads(contract_file.read_text())
    if contract["provenance"]["sceneSha256"] != sha(geometry_dir / "scene.json"):
        raise RuntimeError("Walking collision contract has a stale source manifest")
    if contract["provenance"]["sourceObjSha256"] != scene["objSha256"]:
        raise RuntimeError("Walking collision contract has a stale source OBJ")
    if contract["provenance"]["writerSha256"] != sha(ROOT / "scripts/unreal/walking.mjs"):
        raise RuntimeError("Walking collision exporter changed")
    actor_system = unreal.get_editor_subsystem(unreal.EditorActorSubsystem)
    references = {}
    for actor in actor_system.get_all_level_actors():
        ids = [str(tag) for tag in actor.get_editor_property("tags") if str(tag) in meshes_by_id]
        if not ids:
            continue
        if len(ids) != 1 or ids[0] in references:
            raise RuntimeError("Walking canonical actor identity is ambiguous")
        components = [component for component in actor.get_components_by_class(unreal.StaticMeshComponent)
                      if component.get_editor_property("static_mesh") == meshes_by_id[ids[0]]]
        if len(components) != 1:
            raise RuntimeError("Walking canonical mesh component is ambiguous: " + ids[0])
        references[ids[0]] = (actor, components[0])
    required = {entry["objectId"] for key in ("walkSurfaces", "staticBlockers", "capturedClosedBlockers")
                for entry in contract[key]}
    if not required.issubset(references):
        raise RuntimeError("Walking contract refers to missing canonical actors")

    changed_meshes, supports, closed, static = {}, [], [], []

    def collision(component):
        mesh = component.get_editor_property("static_mesh")
        body = mesh.get_editor_property("body_setup")
        if not body:
            raise RuntimeError("Walking mesh lacks BodySetup")
        body.set_editor_property("never_needs_cooked_collision_data", False)
        body.set_editor_property("collision_trace_flag", unreal.CollisionTraceFlag.CTF_USE_COMPLEX_AS_SIMPLE)
        component.set_collision_profile_name("BlockAll")
        component.set_collision_enabled(unreal.CollisionEnabled.QUERY_AND_PHYSICS)
        component.set_mobility(unreal.ComponentMobility.STATIC)
        changed_meshes[mesh.get_path_name()] = mesh

    before = {id_: bounds(references[id_][1]) for id_ in required}
    for entry in contract["staticBlockers"]:
        actor, component = references[entry["objectId"]]
        collision(component)
        static.append({"objectId": entry["objectId"], "component": component.get_path_name(),
                       "collision": "source-triangles-complex-as-simple"})
    for entry in contract["capturedClosedBlockers"]:
        actor, component = references[entry["objectId"]]
        collision(component)
        for obj in (actor, component):
            add_tags(obj, entry["runtimeTags"])
        closed.append({"objectId": entry["objectId"], "doorId": entry["doorId"],
                       "actor": actor.get_path_name(), "component": component.get_path_name(),
                       "state": "captured-closed", "collision": "source-triangles-complex-as-simple"})

    for entry in contract["walkSurfaces"]:
        actor, component = references[entry["objectId"]]
        offset = entry["supportOffsetCm"]
        if offset != 0:
            source_component = component
            support = actor_system.spawn_actor_from_class(unreal.StaticMeshActor, unreal.Vector(), unreal.Rotator())
            if not support:
                raise RuntimeError("Could not create exact translated walking support")
            support.set_actor_label("Podklad chôdze · " + entry["walkSurfaceId"])
            support.set_folder_path("Brezi/WalkingSupport")
            add_tags(support, ["BreziGenerated", *entry["runtimeTags"], *entry["supportRuntimeTags"]])
            component = support.get_component_by_class(unreal.StaticMeshComponent)
            component.set_mobility(unreal.ComponentMobility.MOVABLE)
            component.set_static_mesh(source_component.get_editor_property("static_mesh"))
            transform = source_component.get_world_transform()
            position = transform.translation
            position.z += offset
            transform.translation = position
            # EditorActorSubsystem relies on Typed Elements, which are absent
            # in this commandlet. Use the actor API, then verify actual bounds.
            support.set_actor_transform(transform, False, True)
            component.set_visibility(False)
            component.set_hidden_in_game(True)
            component.set_cast_shadow(False)
            component.set_visible_in_ray_tracing(False)
            component.set_affect_distance_field_lighting(False)
            add_tags(component, [*entry["runtimeTags"], *entry["supportRuntimeTags"]])
            add_tags(actor, ["BreziWalkVisualSource"])
            add_tags(source_component, ["BreziWalkVisualSource"])
            actual = bounds(component)
            expected = {key: [value + (offset if axis == 2 else 0)
                              for axis, value in enumerate(before[entry["objectId"]][key])]
                        for key in ("min", "max")}
            error = max(abs(actual[key][axis] - expected[key][axis])
                        for key in ("min", "max") for axis in range(3))
            if error > 0.001:
                raise RuntimeError("Walking support differs from its authored translation")
            supports.append({"objectId": entry["objectId"], "walkSurfaceId": entry["walkSurfaceId"],
                             "actor": support.get_path_name(), "component": component.get_path_name(),
                             "supportOffsetCm": offset, "nativeBoundsCm": actual,
                             "maxTranslationErrorCm": error, "visible": False,
                             "sourceMeshReused": component.get_editor_property("static_mesh").get_path_name()})
        else:
            for obj in (actor, component):
                add_tags(obj, entry["runtimeTags"])
        collision(component)

    for id_ in required:
        if bounds(references[id_][1]) != before[id_]:
            raise RuntimeError("Walking authoring changed canonical visual geometry: " + id_)
    for mesh in changed_meshes.values():
        if not unreal.EditorAssetLibrary.save_loaded_asset(mesh):
            raise RuntimeError("Walking collision mesh did not save")
    return {"status": "walking-collision-authored", "contractSha256": sha(contract_file),
            "surfaceCount": len(contract["walkSurfaces"]), "translatedSupportCount": len(supports),
            "staticBlockerCount": len(static), "staticBlockers": static,
            "capturedClosedBlockerCount": len(closed), "supports": supports, "closedBlockers": closed,
            "canonicalVisualBoundsUnchanged": True,
            "assetHashes": _asset_hashes(changed_meshes),
            "pipelineFiles": {OWNER: sha(Path(__file__)), "scripts/unreal/walking.mjs": sha(ROOT / "scripts/unreal/walking.mjs")},
            "verification": {"editorAssetsAuthored": True, "runtimeQueriesVerified": False,
                             "characterMovementVerified": False, "doorAnimationImplemented": False}}
